python/cudf_polars/cudf_polars/experimental/scheduler.py

```diff
@@ -0,0 +1,171 @@
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+    from typing import TypeAlias
+
+
+Key: TypeAlias = str | tuple[str, Unpack[tuple[int, ...]]]
+Graph: TypeAlias = MutableMapping[Key, Any]
+T_ = TypeVar("T_")
+
+
+# NOTE: This is a slimmed-down version of the single-threaded
+# (synchronous) scheduler in `dask.core`.
+#
+# Key Differences:
+# * We do not allow a task to contain a list of key names.
+#   Keys must be distinct elements of the task.
+# * We do not support nested tasks.
+
+
+def istask(x: Any) -> bool:
+    """Check if x is a callable task."""
+    return isinstance(x, tuple) and bool(x) and callable(x[0])
+
+
+def is_hashable(x: Any) -> bool:
+    """Check if x is hashable."""
+    try:
+        hash(x)
+    except TypeError:
+        return False
+    else:
+        return True
+
+
+def _execute_task(arg: Any, cache: Mapping) -> Any:
+    """Execute a compute task."""
+    if istask(arg):
+        return arg[0](*(_execute_task(a, cache) for a in arg[1:]))
+    elif is_hashable(arg):
+        return cache.get(arg, arg)
+    else:
+        return arg
+
+
+def required_keys(key: Key, graph: Graph) -> list[Key]:
+    """
+    Return the dependencies to extract a key from the graph.
+
+    Parameters
+    ----------
+    key
+        Root key we want to extract.
+    graph
+        The full task graph.
+
+    Returns
+    -------
+    List of other keys needed to extract ``key``.
+    """
+    maybe_task = graph[key]
+    return [
+        k
+        for k in (
+            maybe_task[1:]
+            if istask(maybe_task)
+            else [maybe_task]  # maybe_task might be a key
+        )
+        if is_hashable(k) and k in graph
+    ]
+
+
+def toposort(graph: Graph, dependencies: Mapping[Key, list[Key]]) -> list[Key]:
+    """Return a list of task keys sorted in topological order."""
+    # Stack-based depth-first search traversal. This is based on Tarjan's
+    # method for topological sorting (see wikipedia for pseudocode)
+    ordered: list[Key] = []
+    completed: set[Key] = set()
+    seen: set[Key] = set()
+
+    for key in graph:
+        if key in completed:
+            continue
+        nodes: list[Key] = [key]
+        while nodes:
+            # Keep current node on the stack until all descendants are visited
+            current = nodes[-1]
+            if current in completed:  # pragma: no cover
+                # Already fully traversed descendants of current
+                nodes.pop()
+                continue
+            seen.add(current)
+
+            # Add direct descendants of current to nodes stack
+            next_nodes = []
+            for nxt in dependencies[current]:
+                if nxt not in completed:
+                    if nxt in seen:  # pragma: no cover
+                        # Cycle detected!
+                        raise RuntimeError(
+                            f"Cycle detected in the task graph! Key: {nxt}"
+                        )
```

Comment:
In practice will this ever occur as user error, or would it always be a bug in our own code? If the latter, once we're confident in the implementation we could potentially remove this and speed up this whole loop by just writing `next_nodes.extend(set(dependencies[current]) - completed)` or similar (again, assuming that dependency ordering need not be preserved).

Reply:
Must be a bug in our code, it's not possible to construct a cyclic query plan.

Reply:
The cycle checks are in the Dask-scheduler code, but they need to support arbitrary user-defined graphs. In cudf-polars, we are always constructing the graph ourselves. Therefore, I think it's fine to remove `seen` (and cycle detection altogether).

Reply:
I'm fine pushing such optimizations to a follow-up since it sounds like the plan is to get this in and then make improvements later.